File: python/백준/그래프/밸만-포드/1865 웜홀.py
import sys
input = sys.stdin.readline

TC = int(input())

for _ in range(TC):
    N, M, W = map(int, input().split())
    
    edges = []
    
    # 도로 (양방향)
    for _ in range(M):
        S, E, T = map(int, input().split())
        edges.append((S, E, T))
        edges.append((E, S, T))
    
    # 웜홀 (단방향, 음수)
    for _ in range(W):
        S, E, T = map(int, input().split())
        edges.append((S, E, -T))
    
    dist = [0] * (N + 1)  # ⭐ 핵심
    
    has_cycle = False
    
    for i in range(N):
        for u, v, w in edges:
            if dist[v] > dist[u] + w:
                dist[v] = dist[u] + w
                
                if i == N - 1:
                    has_cycle = True
                    break
        
        if has_cycle:
            break
    
    print("YES" if has_cycle else "NO")

File: cpp/백준/그래프/밸만-포드/1865 웜홀.cpp
#include <iostream>
#include <vector>
#include <tuple>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int TC;
    cin >> TC;
    while (TC--) {
        int N, M, W;
        cin >> N >> M >> W;

        vector<tuple<int, int, int>> edges;

        for (int i = 0; i < M; ++i) {
            int S, E, T;
            cin >> S >> E >> T;
            edges.push_back({S, E, T});
            edges.push_back({E, S, T});
        }

        for (int i = 0; i < W; ++i) {
            int S, E, T;
            cin >> S >> E >> T;
            edges.push_back({S, E, -T});
        }

        vector<long long> dist(N + 1, 0);
        bool has_cycle = false;

        for (int i = 0; i < N; ++i) {
            for (auto [u, v, w] : edges) {
                if (dist[v] > dist[u] + w) {
                    dist[v] = dist[u] + w;
                    if (i == N - 1) {
                        has_cycle = true;
                        break;
                    }
                }
            }
            if (has_cycle) break;
        }

        cout << (has_cycle ? "YES" : "NO") << "\n";
    }
    return 0;
}
